sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuUnboundedToUnboundedAggWindowExec.scala

```diff
@@ -0,0 +1,1172 @@
+    withResource(GpuColumnVector.from(inputCB)) { inputTable =>
+      withResource(cudf.Scalar.fromInt(1)) { one =>
+        withResource(
+          cudf.ColumnVector.fromScalar(one,
+            inputTable.getColumn(0).getRowCount.asInstanceOf[Int])) { ones =>
+          val columns = Range(0, inputTable.getNumberOfColumns)
+            .map {inputTable.getColumn} :+ ones
+          withResource(new cudf.Table(columns: _*)) { preProcessedTable =>
+            GpuColumnVector.from(preProcessedTable,
+              boundStages.inputTypes.toArray :+ IntegerType)
+          }
+        }
+      }
+    }
+  }
+
+  private def groupByAggregate(inputCB: ColumnarBatch) = {
+    // Note: The data is always ordered first by the grouping keys,
+    // as ASC NULLS FIRST, regardless of how the order-by columns
+    // are ordered.  This happens prior to the window exec, since
+    // the GpuSortOrder is upstream from the window exec.
+    val groupByOptions = cudf.GroupByOptions.builder()
+                             .withIgnoreNullKeys(false)
+                             .withKeysSorted(true)
+                             .build()
+    val cudfAggregates = boundStages.cudfUpdateAggregates
+    val aggInputOrdinals = boundStages.aggInputOrdinals
+    val cudfAggsOnColumns = cudfAggregates.zip(aggInputOrdinals).map {
+      case (cudfAgg, ord) => cudfAgg.groupByAggregate.onColumn(ord)
+    }
+
+    withResource(GpuColumnVector.from(inputCB)) { inputTable =>
+      val aggResults = inputTable.groupBy(groupByOptions, boundStages.groupColumnOrdinals: _*)
+        .aggregate(cudfAggsOnColumns: _*)
+      // The COUNT aggregate result (at the end) is returned from libcudf as an INT32,
+      // while Spark expects an `INT64`.  This needs to be scaled up.
+      withResource(aggResults) {
+        upscaleCountResults
+      }
+    }
+  }
+
+  override def next(): FirstPassAggResult = {
+    if (!hasNext) {
+      throw new NoSuchElementException()
+    }
+    if (subIterator.exists(_.hasNext)) {
+      subIterator.map(_.next()).get
+    } else {
+      val currIter = withRetry(getSpillableInputBatch, splitSpillableInHalfByRows) { scb =>
+        withResource(scb.getColumnarBatch()) { cb =>
+          withResource(preProcess(cb)) { preProcessedInput =>
+            withResource(groupByAggregate(preProcessedInput)) { aggResultTable =>
+              val rideAlongColumns = GpuProjectExec.project(preProcessedInput,
+                boundStages.boundRideAlong)
+
+              FirstPassAggResult(
+                adoptAndMakeSpillable(rideAlongColumns),
+                toSpillableBatch(aggResultTable,
+                  boundStages.groupingColumnTypes ++ boundStages.aggResultTypes))
+            }
+          }
+        }
+      }
+      val result = currIter.next()
+      subIterator = Some(currIter)
+      result
+    }
+  }
+}
+
+case class PartitionedFirstPassAggResult(firstPassAggResult: FirstPassAggResult,
```

Comment:
nit: I don't think this really fits with being a `case class`. We might want to change it to be just a class.

Reply:
Thanks for pointing that out, @revans2.  That helped me better understand the use of `case class`.  I've changed this to a regular class now.